[VAOS] Calendar Widget

narin opened the pull request:
## Description
PR for calendar widget that will be used for direct schedule and appointment request.  This PR is for the calendar widget only.  Connecting it to the new appointment flow will be handled in a different PR (https://github.com/department-of-veterans-affairs/vets-website/pull/10941).

## Testing done
Local and unit tests

## Screenshots

### Direct Schedule
![Screen Shot 2019-10-10 at 11 00 37 PM](https://user-images.githubusercontent.com/786704/66628403-9734bd00-ebb3-11e9-9120-e854e0b7c546.png)

### Request
![image](https://user-images.githubusercontent.com/786704/66786461-3e0ba880-ee95-11e9-83ff-ec7b44fb7f61.png)


## Acceptance criteria
- [x] Calendar only shows available dates as clickable
- [x] Calendar shows unavailable dates as greyed out
- [x] In direct scheduling, once date selected, available times expand beneath
- [x] In request flow, once date is selected, am / pm option expands beneath (confirm this behavior with @rtwell )
- [x] Calendar shows dates and times in facility timezone (confirm this as well)
- [x] Calendar shows available dates up to 45 days in advance (this specific number may change)
- [x] Calendar shows available dates for two stacked months at a time
- [x] Calendar includes 'next' / 'previous' buttons that allow veterans to advance months
- [x] Calendar looks like [current mocks]

## Definition of done
- [ ] Events are logged appropriately
- [ ] Documentation has been updated, if applicable
- [ ] A link has been provided to the originating GitHub issue (or connected to it via ZenHub)
- [ ] No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs

narin committed:
Calendar widget and additional options

narin committed:
Style options and selected checkbox

narin committed:
Rename some variables and move some functiont to calendar utils

narin committed:
Update for direct sechduling

narin committed:
Disable days before today

narin committed:
Add validation for multiselct

narin committed:
Refactor cell and options into stateless components

narin committed:
Add helper functions to convert array to obj and visa versa

narin committed:
Clean up and minor refactoring

narin committed:
Fix comment typo

narin committed:
Replace some scss with utility classes

narin commented on src/applications/vaos/containers/DateTimeSelectPage.jsx:
```diff
@@ -0,0 +1,120 @@
+
+const pageKey = 'dateTimeSelect';
+
+const institutionTimezone = 'America/Denver';
```
Temporarily hardcoded this timezone.  Will be grabbing this from selected institution/clinic.

narin committed:
Rename function paramter

narin commented on src/applications/vaos/containers/DateTimeSelectPage.jsx:
```diff
@@ -0,0 +1,120 @@
+}
+
+function mapStateToProps(state) {
+  return { ...getFormPageInfo(state, pageKey), availableSlots: slots };
```
`availableSlots` will come from redux state instead of being hardcoded like this.

jbalboni commented:
@narin Looks good so far. This is what I've noticed from playing around with it:

1. I'm not sure if this is properly called out in the ticket, but for requested dates, the AM/PM choice should be checkboxes, so that you can select both at once. This could be something we change in a later ticket, but Ryan and I had a discussion about it and since you can choose the same two dates with AM and PM on the current app, we need to allow that here.
2. The hidden buttons need `aria-hidden` on them, since they're still in the markup.
3. The buttons probably need `aria-label` attributes that have a more descriptive label than just the day of the month that's in the button text.
4. I'd wait for @1Copenut to weigh in on this before changing anything, but I suspect we want the tab order to from the day you've currently selected, to the time radio buttons, then to the next day. Currently, you have to tab through all the days in single row to get to the time radio buttons.
5. We might want `aria-expanded` to be set to true on buttons that are selected and open. This is sort of related to item 4, so Trevor would know for sure.
6. Not sure if you've looked into it, but we might want to figure out how to make sure we're not pulling in more than what we need from moment-timezone. It looks like the size of it varies depending on what you need.
7. I wonder if we want to disable the Next button once we've hit the end of the available appointment slots or at some kind of limit for requested dates? I think there's a max we can schedule out.

narin commented:
@jbalboni All great points. Thanks for taking a look.

narin committed:
Renamed var for clarity

narin committed:
Set max month to display based on available dates or default setting of 45 days

narin committed:
Add aria attributes

narin committed:
Refactored to allow for checkboxes.  Use array to keep track of selected state instead of map.

narin committed:
Remove unused utility functions

narin commented:
Hey @jbalboni I've addressed all your suggestions aside from the `moment-timezone` one.  The NPM package seems to be an all or nothing type situation. Working on finding a more lightweight replacement for it.

In regards to the AM/PM checkboxes, initially I was using an object to keep track of selected dates, so that I could easily look up `selectedDates[date]`.  This worked fine when there was only 1 possible additional option (radio buttons). However with the checkbox requirement for AM/PM, I modified the code to use an array to store the selected combinations.  This required some refactoring and the addition of a couple helper functions.

Please take a look and review if the UI portion of this looks good so we can get it merged and @1Copenut can start taking a look.  Will create a separate PR to hook up to the SchemaForm/new appointment flow. Thanks.

narin committed:
Fix typo

narin committed:
Create form fields for calendar

1Copenut reviewed (commented):
I posted some 508 items and a Codepen mockup in the following issue comment: https://github.com/department-of-veterans-affairs/va.gov-team/issues/2569#issuecomment-544709696

1Copenut commented:
@narin I talked with @markgreenburg this afternoon. I'd say if you can get the `roles` added as part of this PR, let's do that, and I'll write specific issue tickets for the axe and keyboard items under the https://app.zenhub.com/workspaces/vsp-5cedc9cce6e3335dc5a49fc4/issues/department-of-veterans-affairs/va.gov-team/2671 epic.

narin committed:
Styling fixes

narin committed:
Integrate widget into form

narin committed:
Mobile styling fixes

narin committed:
Implemented several of Jeffs suggestions and removed unrelated changes

narin committed:
Undo changes to openFacilityPage()

narin committed:
Split PR into 2

narin committed:
Split PR into 2

narin committed:
Undo addition of moment-timezone

narin committed:
Undo addition of moment-timezone

narin committed:
Add unit test for <CalendarWeekdayHeader/>

narin committed:
Add unit test for <CalendarNavigation/>

narin committed:
Add unit tests for <CalendarCheckboxOption /> and <CalendarRadioOption />

narin committed:
Add unit test for <CalendarCell />

narin committed:
Add unit test for CalendarRow

narin committed:
Add unit test for <CalendarWidget />

narin committed:
Fix unit test for CalendarNavigation

narin committed:
Add additional tests for <CalendarCell />
